gcloud/connection.py

```diff
@@ -91,14 +95,23 @@ def credentials(self):
     def http(self):
         """A getter for the HTTP transport used in talking to the API.
 
-        :rtype: :class:`httplib2.Http`
-        :returns: A Http object used to transport data.
+        This will return a thread-local :class:`httplib2.Http` instance unless
+        a custom transport has been provided to the `Connection` constructor.
+
+        :rtype: :class:`httplib2.Http` or the custom HTTP transport specifed
+            to the connection constructor.
+        :returns: An `Http` object used to transport data.
         """
-        if self._http is None:
-            self._http = httplib2.Http()
+        if self._http is not None:
+            return self._http
+
+        if not hasattr(self._local, 'http'):
+            self._local.http = httplib2.Http()
             if self._credentials:
-                self._http = self._credentials.authorize(self._http)
-        return self._http
+                self._local.http = self._credentials.authorize(
```

Comment:
Agreed, however the `credentials` property on Connection is read-only, so it should only be changed/set during construction.

If you feel it's worthwhile I can switch to just:

```
try:
    self._local.http = self._credentials.authorize(self._local.http)
except AttributeError:
    pass
```

or somehow add a lock.